### PR TITLE
deneb - validator beacon rest apis

### LINKED ISSUE #12248: EIP4844 Feature Completion Checklist
This issue tracks the remaining tasks for completing the EIP4844 feature on Deneb. The following tasks have already been accomplished:

- State transition implementation is up to spec. Latest spec tests passed
- Validator signing blobs and broadcasting blobs to individual gossip net are working
- DB pruning for blobs is working
- Most of current spec'ed beacon API endpoints are implemented and tested
- Blob gossip is working
- Blob by range syncing is working
- Blob by root syncing is working
- A node can sync from Genesis to post Deneb on the block with blobs
- Prysmctl tool to request blobs by range

The following tasks are remaining:
- [ ] Implement backfilling for blocks and blobs
- [ ] Implement builder path for Deneb blind block and blind blobs 
   - [x]  update builder client handling https://github.com/ethereum/builder-specs/pull/61
- [ ] Implement a better heuristic approach to request missing blob
- [ ] Implement blob storage separation. Users can choose to store blobs in other file systems if they wish to
- [ ] Add and optimize sidecar KZG proof checks
- [ ] Handle the edge case where the finalized epoch is older than the data retention epoch
- [ ] Clean up the initial sync code. Perhaps using this excellent opportunity to rewrite the majority of the initial sync code
- [ ] Test thoroughly and refine the user experience
- [x] Update Beacon API endpoints:
  - [x] `getStateV2` json
  - [x] `getStateV2` ssz
  - [x] `getBlockV2` json
  - [x] `getBlockV2` ssz
  - [x] `getBlindedBlock` json
  - [x] `getBlindedBlock` ssz
  - [x] `publishBlock` json
  - [x] `publishBlock` ssz 
  - [x] `publishBlockV2` json
  - [x] `publishBlockV2` ssz
  - [x] `publishBlindedBlock` json
  - [x] `publishBlindedBlock` ssz 
  - [x] `publishBlindedBlockV2` json
  - [x] `publishBlindedBlockV2` ssz 
  - [x] `produceBlindedBlock` json
  - [x] `produceBlindedBlock` ssz 
  - [x]  `produceBlockV2` json
  - [x] `produceBlockV2` ssz 
  - [x]  `produceBlockV3` json https://github.com/ethereum/beacon-APIs/pull/339
  - [x]  `produceBlockV3` ssz
  - [x]  `getBlobs` json
  - [x]  `getBlobs` ssz
 - [ ] add Web3signer support for Deneb blocks
 - [ ] add Web3signer support for signed blobs
 - [ ] update validator rest api calls to support deneb:
   - [ ] `getBlock`
   - [ ] `streamBlock`
   - [ ] `submitBlock`
   - [ ] `submitBlindedBlock`
 - [ ] E2E for Deneb

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

adds deneb support for validator rest apis like submit block and get block to support deneb. we are not fully utilizing these endpoints until we do some further performance testing but this will make sure that minimal scenario e2e does not break.

**Which issues(s) does this PR fix?**

Fixes #12248

### CODE REVIEW

File: validator/client/beacon-api/get_beacon_block.go [hunk 1]
```diff
@@ -98,7 +99,16 @@ func (c beaconApiValidatorClient) getBeaconBlock(ctx context.Context, slot primi
...
 		}
-
+	case "deneb":
+		jsonDenebBlockContents := shared.BeaconBlockContentsDeneb{}
+		if err := decoder.Decode(&jsonDenebBlockContents); err != nil {
+			return nil, errors.Wrap(err, "failed to decode deneb block response json")
+		}
+		genericBlock, err := jsonDenebBlockContents.ToGeneric()
+		if err != nil {
+			return nil, errors.Wrap(err, "could not convert deneb blockcontents to generic")
```
Comment (rkapka) [inline validator/client/beacon-api/get_beacon_block.go:109]:
```suggestion
			return nil, errors.Wrap(err, "could not convert deneb block contents to generic block")
```

File: validator/client/beacon-api/propose_beacon_block.go [hunk 2]
```diff
@@ -90,6 +91,35 @@ func (c beaconApiValidatorClient) proposeBeaconBlock(ctx context.Context, in *et
...
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to compute block root for deneb beacon block")
+		}
+		signedBlock, err := shared.SignedBeaconBlockContentsDenebFromConsensus(blockType.Deneb)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to convert deneb beacon block content")
+		}
+		marshalledSignedBeaconBlockJson, err = json.Marshal(signedBlock)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to marshall deneb beacon block content")
```
Comment (rkapka) [inline validator/client/beacon-api/propose_beacon_block.go:106]:
```suggestion
			return nil, errors.Wrap(err, "failed to marshal deneb beacon block contents")
```

File: validator/client/beacon-api/propose_beacon_block.go [hunk 3]
```diff
@@ -90,6 +91,35 @@ func (c beaconApiValidatorClient) proposeBeaconBlock(ctx context.Context, in *et
...
 		}
+	case *ethpb.GenericSignedBeaconBlock_Deneb:
+		consensusVersion = "deneb"
+		beaconBlockRoot, err = blockType.Deneb.Block.HashTreeRoot()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to compute block root for deneb beacon block")
+		}
+		signedBlock, err := shared.SignedBeaconBlockContentsDenebFromConsensus(blockType.Deneb)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to convert deneb beacon block content")
```
Comment (rkapka) [inline validator/client/beacon-api/propose_beacon_block.go:102]:
```suggestion
			return nil, errors.Wrap(err, "failed to convert deneb beacon block contents")
```

File: validator/client/beacon-api/propose_beacon_block.go [hunk 4]
```diff
@@ -90,6 +91,35 @@ func (c beaconApiValidatorClient) proposeBeaconBlock(ctx context.Context, in *et
...
+	case *ethpb.GenericSignedBeaconBlock_BlindedDeneb:
+		blinded = true
+		consensusVersion = "deneb"
+		beaconBlockRoot, err = blockType.BlindedDeneb.Block.HashTreeRoot()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to compute block root for blinded deneb beacon block")
+		}
+		signedBlock, err := shared.SignedBlindedBeaconBlockContentsDenebFromConsensus(blockType.BlindedDeneb)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to convert blinded deneb beacon block content")
```
Comment (rkapka) [inline validator/client/beacon-api/propose_beacon_block.go:117]:
```suggestion
			return nil, errors.Wrap(err, "failed to convert blinded deneb beacon block contents")
```

File: validator/client/beacon-api/propose_beacon_block.go [hunk 5]
```diff
@@ -90,6 +91,35 @@ func (c beaconApiValidatorClient) proposeBeaconBlock(ctx context.Context, in *et
...
+		}
+		marshalledSignedBeaconBlockJson, err = json.Marshal(signedBlock)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to marshall blinded deneb beacon block content")
```
Comment (rkapka) [inline validator/client/beacon-api/propose_beacon_block.go:121]:
```suggestion
			return nil, errors.Wrap(err, "failed to marshal blinded deneb beacon block contents")
```

**Review by rkapka — APPROVED** [review]
